### PR TITLE
Toggle support for `QStartNoAckMode`

### LINKED ISSUE #134: Make `QStartNoAckMode` optional
I want to use IDA Pro as a GDB client to connect the example program in my project [udbserver](https://github.com/bet4it/udbserver):
```bash
$ cargo run --example server
```
IDA Pro asked me the PID to attach to, and failed to attach after I provided a pid:
```
 TRACE gdbstub::protocol::recv_packet > <-- +
 TRACE gdbstub::protocol::recv_packet > <-- $qSupported:xmlRegisters=i386,arm,mips#12
 TRACE gdbstub::protocol::response_writer > --> $PacketSize=1000;vContSupported+;multiprocess+;QStartNoAckMode+;swbreak+;hwbreak+;qXfer:features:read+#47
 TRACE gdbstub::protocol::recv_packet     > <-- +
 TRACE gdbstub::protocol::recv_packet     > <-- $QStartNoAckMode#b0
 TRACE gdbstub::protocol::response_writer > --> $OK#9a
 TRACE gdbstub::protocol::recv_packet     > <-- $!#21
 INFO  gdbstub::stub::core_impl           > Unknown command: Ok("!")
 TRACE gdbstub::protocol::response_writer > --> $#00
 TRACE gdbstub::protocol::recv_packet     > <-- $Hg0#df
 TRACE gdbstub::protocol::response_writer > --> $OK#9a
 TRACE gdbstub::protocol::recv_packet     > <-- $qXfer:features:read:target.xml:0,fff#7d
 TRACE gdbstub::protocol::response_writer > --> $m
        <target version="1.0">
            <architecture>arm</architecture>
            <feature name="org.gnu.gdb.arm.core">
              <reg name="r0" bitsize="32" type="uint32"/>
              <reg name="r1" bitsize="32" type="uint32"/>
              <reg name="r2" bitsize="32" type="uint32"/>
              <reg name="r3" bitsize="32" type="uint32"/>
              <reg name="r4" bitsize="32" type="uint32"/>
              <reg name="r5" bitsize="32" type="uint32"/>
              <reg name="r6" bitsize="32" type="uint32"/>
              <reg name="r7" bitsize="32" type="uint32"/>
              <reg name="r8" bitsize="32" type="uint32"/>
              <reg name="r9" bitsize="32" type="uint32"/>
              <reg name="r10" bitsize="32" type="uint32"/>
              <reg name="r11" bitsize="32" type="uint32"/>
              <reg name="r12" bitsize="32" type="uint32"/>
              <reg name="sp" bitsize="32" type="data_ptr"/>
              <reg name="lr" bitsize="32"/>
              <reg name="pc" bitsize="32" type="code_ptr"/>
              <reg name="cpsr" bitsize="32"/>
            </feature>
        </target>
    #5f
 TRACE gdbstub::protocol::recv_packet     > <-- $?#3f
 TRACE gdbstub::protocol::response_writer > --> $T05thread:01;#07
 TRACE gdbstub::protocol::recv_packet     > <-- $vAttach;1#37
 INFO  gdbstub::stub::core_impl           > Unknown command: Ok("vAttach;1")
 TRACE gdbstub::protocol::response_writer > --> $#00
```
But if `QStartNoAckMode` is disabled, IDA Pro won't ask for the PID and everything works well:
```
 TRACE gdbstub::protocol::recv_packet > <-- +
 TRACE gdbstub::protocol::recv_packet > <-- $qSupported:xmlRegisters=i386,arm,mips#12
 TRACE gdbstub::protocol::response_writer > --> $PacketSize=1000;vContSupported+;multiprocess+;swbreak+;hwbreak+;qXfer:features:read+#31
 TRACE gdbstub::protocol::recv_packet     > <-- +
 TRACE gdbstub::protocol::recv_packet     > <-- $!#21
 INFO  gdbstub::stub::core_impl           > Unknown command: Ok("!")
 TRACE gdbstub::protocol::response_writer > --> $#00
 TRACE gdbstub::protocol::recv_packet     > <-- +
 TRACE gdbstub::protocol::recv_packet     > <-- $Hg0#df
 TRACE gdbstub::protocol::response_writer > --> $OK#9a
 TRACE gdbstub::protocol::recv_packet     > <-- +
 TRACE gdbstub::protocol::recv_packet     > <-- $qXfer:features:read:target.xml:0,fff#7d
 TRACE gdbstub::protocol::response_writer > --> $m
        <target version="1.0">
            <architecture>arm</architecture>
            <feature name="org.gnu.gdb.arm.core">
              <reg name="r0" bitsize="32" type="uint32"/>
              <reg name="r1" bitsize="32" type="uint32"/>
              <reg name="r2" bitsize="32" type="uint32"/>
              <reg name="r3" bitsize="32" type="uint32"/>
              <reg name="r4" bitsize="32" type="uint32"/>
              <reg name="r5" bitsize="32" type="uint32"/>
              <reg name="r6" bitsize="32" type="uint32"/>
              <reg name="r7" bitsize="32" type="uint32"/>
              <reg name="r8" bitsize="32" type="uint32"/>
              <reg name="r9" bitsize="32" type="uint32"/>
              <reg name="r10" bitsize="32" type="uint32"/>
              <reg name="r11" bitsize="32" type="uint32"/>
              <reg name="r12" bitsize="32" type="uint32"/>
              <reg name="sp" bitsize="32" type="data_ptr"/>
              <reg name="lr" bitsize="32"/>
              <reg name="pc" bitsize="32" type="code_ptr"/>
              <reg name="cpsr" bitsize="32"/>
            </feature>
        </target>
    #5f
 TRACE gdbstub::protocol::recv_packet     > <-- $?#3f
 TRACE gdbstub::protocol::response_writer > --> $T05thread:01;#07
 TRACE gdbstub::protocol::recv_packet     > <-- +
 TRACE gdbstub::protocol::recv_packet     > <-- $?#3f
 TRACE gdbstub::protocol::response_writer > --> $T05thread:01;#07
 TRACE gdbstub::protocol::recv_packet     > <-- +
 TRACE gdbstub::protocol::recv_packet     > <-- $qfThreadInfo#bb
 TRACE gdbstub::protocol::response_writer > --> $m01#ce
 TRACE gdbstub::protocol::recv_packet     > <-- +
 TRACE gdbstub::protocol::recv_packet     > <-- $qsThreadInfo#c8
 TRACE gdbstub::protocol::response_writer > --> $l#6c
 TRACE gdbstub::protocol::recv_packet     > <-- +
 TRACE gdbstub::protocol::recv_packet     > <-- $g#67
 TRACE gdbstub::protocol::response_writer > --> $00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000100000d3010040#ce
 TRACE gdbstub::protocol::recv_packet     > <-- +
 TRACE gdbstub::protocol::recv_packet     > <-- $qXfer:features:read:target.xml:0,fff#7d
 TRACE gdbstub::protocol::response_writer > --> $m
        <target version="1.0">
            <architecture>arm</architecture>
            <feature name="org.gnu.gdb.arm.core">
              <reg name="r0" bitsize="32" type="uint32"/>
              <reg name="r1" bitsize="32" type="uint32"/>
              <reg name="r2" bitsize="32" type="uint32"/>
              <reg name="r3" bitsize="32" type="uint32"/>
              <reg name="r4" bitsize="32" type="uint32"/>
              <reg name="r5" bitsize="32" type="uint32"/>
              <reg name="r6" bitsize="32" type="uint32"/>
              <reg name="r7" bitsize="32" type="uint32"/>
              <reg name="r8" bitsize="32" type="uint32"/>
              <reg name="r9" bitsize="32" type="uint32"/>
              <reg name="r10" bitsize="32" type="uint32"/>
              <reg name="r11" bitsize="32" type="uint32"/>
              <reg name="r12" bitsize="32" type="uint32"/>
              <reg name="sp" bitsize="32" type="data_ptr"/>
              <reg name="lr" bitsize="32"/>
              <reg name="pc" bitsize="32" type="code_ptr"/>
              <reg name="cpsr" bitsize="32"/>
            </feature>
        </target>
    #5f
 TRACE gdbstub::protocol::recv_packet     > <-- $pf#d6
 TRACE gdbstub::protocol::response_writer > --> $00100000#0c
 TRACE gdbstub::protocol::recv_packet     > <-- +
 TRACE gdbstub::protocol::recv_packet     > <-- $g#67
 TRACE gdbstub::protocol::response_writer > --> $00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000100000d3010040#ce
 TRACE gdbstub::protocol::recv_packet     > <-- +
```
Should we make `QStartNoAckMode` optional to support IDA Pro? Do we have other ways to solve it?

### PR DESCRIPTION
**Pull request by bet4it**
### Description

This PR makes the feature `QStartNoAckMode` toggleable.

Closes #134

### API Stability

- [x] This PR does not require a breaking API change

### Checklist

- Documentation
  - [x] Ensured any public-facing `rustdoc` formatting looks good (via `cargo doc`)
  - [ ] (if appropriate) Added feature to "Debugging Features" in README.md
- Validation
  - [ ] Included output of running `examples/armv4t` with `RUST_LOG=trace` + any relevant GDB output under the "Validation" section below
  - [x] Included output of running `./example_no_std/check_size.sh` before/after changes under the "Validation" section below
- _If implementing a new protocol extension IDET_
  - [ ] Included a basic sample implementation in `examples/armv4t`
  - [x] IDET can be optimized out (confirmed via `./example_no_std/check_size.sh`)
  - [ ] **OR** implementation requires introducing non-optional binary bloat (please elaborate under "Description")
- _If upstreaming an `Arch` implementation_
  - [ ] I have tested this code in my project, and to the best of my knowledge, it is working as intended.

### Validation

<details>
<summary>Before/After `./example_no_std/check_size.sh` output</summary>

### Before

```
target/release/gdbstub-nostd  :
section               size    addr
.interp                 28     792
.note.gnu.property      32     824
.note.gnu.build-id      36     856
.note.ABI-tag           32     892
.gnu.hash               28     928
.dynsym                360     960
.dynstr                204    1320
.gnu.version            30    1524
.gnu.version_r          64    1560
.rela.dyn              408    1624
.init                   27    4096
.text                15158    4128
.fini                   13   19288
.rodata                954   20480
.eh_frame_hdr          244   21436
.eh_frame             1252   21680
.init_array              8   28088
.fini_array              8   28096
.dynamic               432   28104
.got                   136   28536
.data                    8   28672
.bss                     8   28680
.comment                27       0
Total                19497
```

### After 2bc3d4ff1de138d1fa028ecbc2f29aa18d40fc77

```
target/release/gdbstub-nostd  :
section               size    addr
.interp                 28     792
.note.gnu.property      32     824
.note.gnu.build-id      36     856
.note.ABI-tag           32     892
.gnu.hash               28     928
.dynsym                360     960
.dynstr                204    1320
.gnu.version            30    1524
.gnu.version_r          64    1560
.rela.dyn              408    1624
.init                   27    4096
.text                15158    4128
.fini                   13   19288
.rodata                930   20480
.eh_frame_hdr          244   21412
.eh_frame             1252   21656
.init_array              8   28088
.fini_array              8   28096
.dynamic               432   28104
.got                   136   28536
.data                    8   28672
.bss                     8   28680
.comment                27       0
Total                19473
```

### After 3e7e6a8a71b9af51ca99a463e0d9c4c2fe6ce141

```
target/release/gdbstub-nostd  :
section               size    addr
.interp                 28     792
.note.gnu.property      32     824
.note.gnu.build-id      36     856
.note.ABI-tag           32     892
.gnu.hash               28     928
.dynsym                360     960
.dynstr                204    1320
.gnu.version            30    1524
.gnu.version_r          64    1560
.rela.dyn              408    1624
.init                   27    4096
.text                14949    4128
.fini                   13   19080
.rodata                930   20480
.eh_frame_hdr          244   21412
.eh_frame             1252   21656
.init_array              8   28088
.fini_array              8   28096
.dynamic               432   28104
.got                   136   28536
.data                    8   28672
.bss                     8   28680
.comment                27       0
Total                19264
```
</details>

### CODE REVIEW

**Comment by daniel5151** [comment]
Good start, but you'll also need to tweak the code in `commands.rs` to compile-out the `QStartNoAckMode` packet parsing if no ack mode is disabled.

Some pointers on that:
- https://github.com/daniel5151/gdbstub/blob/bf4ad90/src/protocol/commands.rs#L249-L251
- https://github.com/daniel5151/gdbstub/blob/bf4ad90/src/protocol/commands.rs#L94

And doing so will have knock-on effects in `stub/core_impl`, whereby you'll need to split `QStartNoAckMode` handling out of `Base` and into its own packet group.

**Comment by daniel5151** [comment]
Yes, to be clear: I'm saying that you should move the `QStartNoAckMode` code out of the `base use 'a` category, and move it into a new `no_ack_mode` category (the same way `X` packet parsing _could_ have been shoved into `base use 'a`, but was instead put into it's own section)

**Comment by bet4it** [comment]
Even if we don't include `QStartNoAckMode` in features string, client can still enable it by `set remote noack-packet on` before connecting, so we could leave the handle code there?

**Comment by daniel5151** [comment]
`gdbstub` is all about "you don't pay for what you don't use".

If the `gdbstub` implementation has decided to _statically_ disable support for noack mode (e.g: by overriding the method to return a static `false` boolean from `use_no_ack_mode`), then the stub shouldn't have to pay the packet parsing cost of `QStartNoAckMode`.

The details of the `use_no_ack_mode` implementation shouldn't be important to the consumer of the library (i.e: the fact that we skip reporting `;QStartNoAckMode+` in the features string). From the `gdbstub` user's perspective, the usage contract should be as simple as: "if I disable support for noack mode, then my stub binary shrinks, in exchange for a slightly slower protocol" (and the docs of the `use_no_ack_mode` method should reflect that fact).

File: src/target/mod.rs [hunk 1]
```diff
@@ -530,6 +530,14 @@ pub trait Target {
         <Self::Arch as Arch>::single_step_gdb_behavior()
     }
 
+    /// Enable/disable `QStartNoAckMode`
```
Comment (daniel5151) [inline src/target/mod.rs:533]: I don't think these docs are up to snuff...
That said, for expediency's sake, I don't mind cleaning these up after merge.

In the future though, docs should avoid mentioning the nitty-gritty details of their underlying implementation whenever possible, and instead, be written from the perspective of user-visible behavior. The point of `gdbstub` is that users really shouldn't care what a `QStartNoAckMode` packet is.

i.e: i'll tweak these docs to paraphrase from https://sourceware.org/gdb/onlinedocs/gdb/Packet-Acknowledgment.html#Packet-Acknowledgment

**Review by daniel5151 — APPROVED** [review]
Nice, this LGTM! (docs outstanding)

I'll merge this in later today, and cut a new release with this feature included.

Thanks again for the contribution!

**Comment by daniel5151** [comment]
gdbstub 0.6.6 has been published to crates.io

**Pull request merged**